src/components/InfiniteScroller/index.js

```diff
@@ -0,0 +1,150 @@
+import React, {PureComponent as Component} from 'react'
+import ReactDOM from 'react-dom'
+import PropTypes from 'prop-types'
+import EventListener from '../EventListener'
+import classNames from '../../utilities/classNames'
+import LoadingDots from '../LoadingDots'
+import { isNodeVisible } from '../../utilities/node'
+import { noop } from '../../utilities/other'
+
+export const propTypes = {
+  className: PropTypes.string,
+  offset: PropTypes.number,
+  isLoading: PropTypes.bool,
+  onLoading: PropTypes.func,
+  onLoaded: PropTypes.func
+}
+const defaultProps = {
+  offset: 0,
+  isLoading: false,
+  onLoading: noop,
+  onLoaded: noop
+}
+
+class InfiniteScroller extends Component {
+  constructor (props) {
+    super()
+    this.state = {
+      isLoading: props.isLoading,
+      nodeScope: window
+    }
+    this.node = null
+    this.handleOnScroll = this.handleOnScroll.bind(this)
+  }
+
+  componentDidMount () {
+    this.setParentNode()
+  }
+
+  componentWillReceiveProps (nextProps) {
+    /* istanbul ignore else */
+    if (nextProps.isLoading !== this.state.isLoading) {
+      this.setState({
+        isLoading: nextProps.isLoading
+      })
+    }
+    /* istanbul ignore else */
+    if (nextProps.isLoading && !this.state.isLoading) {
+      this.handleOnLoaded()
+    }
+  }
+
+  /* istanbul ignore next */
+  // This method is tested, but tested in separate parts.
+  // Unable to validate isNodeVisible in this method. The isNodeVisible
+  // has been thoroughly tested elsewhere.
+  // The handleOnLoading method has been abstracted to be tested in
+  // isolation.
+  handleOnScroll (event) {
+    const { offset } = this.props
+    const { isLoading, nodeScope } = this.state
+
+    if (isLoading ||
+      !isNodeVisible({
+        node: this.node,
+        scope: nodeScope,
+        offset,
+        complete: true
+      })
+    ) return
+
+    this.handleOnLoading()
+  }
+
+  handleOnLoading () {
+    const { onLoading } = this.props
+    this.setState({
+      isLoading: true
+    })
+    onLoading(() => { this.handleOnLoaded() })
+  }
+
+  handleOnLoaded () {
+    const { onLoaded } = this.props
+    onLoaded()
+
+    if (this.state.isLoading) {
+      this.setState({
+        isLoading: false
+      })
+    }
+  }
+
+  setParentNode () {
+    const node = ReactDOM.findDOMNode(this)
+    /* istanbul ignore next */
+    // Tested for node.parentNode, but not for window.
+    // This is a super fail-safe. This will always be parentNode, with the
+    // exception of document or window. Cannot be tested in JSDOM/Enzyme,
+    // since it prohibits mounting on document.body directly.
+    const nodeScope = node && node.parentNode ? node.parentNode : window
```

Comment:
@ItsJonQ Does this mean that the `InfiniteScroll` will only work if it is the direct child of a `Scrollable`? Or rather, if its parent is styled such that it can be scrolled?

Reply:
InfiniteScroller watches scroll for **any** parent container, not just the `Scrollable` component.

And correct, it only triggers on scroll.

I can add 2 new options…

A `scope` prop to watch for scroll of a **specific** node. But I can't imagine this component listening for anything that isn't the parent node.

Also a `loadOnMount`, which does the check when this component is mounted vs only listening for scroll.

Reply:
Sorry, I wasn't clear. My concern is that I've got something like this: 
```<Modal>
  <Scrollable>
     <Conversation>
      <ThreadList>
        <InfiniteScroll />
      </ThreadList>
    </Conversation>
  </Scrollable>
</Modal>
```

In this case, doesn't the `scroll` event fire on the Scrollable? And won't that mean that the `<InfiniteScroll>` will miss it, since it's only listening for its direct parent to have a scroll event?

Reply:
OK so I went ahead and played with this, I have some work locally which add a `infiniteScroller` prop to the `<Modal>` so that I should be able to pass in an InfiniteScroller of my choosing to the Conversation's `<Modal>` in beacon2. I'm really unsure if this is the right approach, though, because I don't like having to be aware of scrolling at the level of the <Modal>? Like, I really want to put the InfiniteScroll down inside my ThreadList or something... 

Seems to me like we can either do what I was working towards, adding an `infiniteScroller` prop to the Modal - or - make the logic which "finds" the parentElement overridable, so that we could pass a custom function in like `<InfiniteScroll getScrollParent={getScrollParent}/>`. 

I'm kinda leaning towards the latter, what do you think? The latter would allow me to pass a `getScrollParent` prop down from the modal into the `ThreadList`, so it would have all the logic it needs for infinite scroll, with an abstract scroll parent that the Modal knows about. What do you think?

Reply:
Updates! I've added a new `scrollParent` prop to `InfiniteScroller`. You can pass any DOM node there, and the component will listen to scroll events on that instead. Otherwise, it will find it's closest parentNode.

I've also adjusted the Modal markup to make it more `<Scrollable>` friendly.

Is it possible for the thing you're building to have this structure?

```
<Modal trigger={myTrigger}>
  <Conversation>
    <ThreadList>
    </ThreadList>
  </Conversation>
  <InfiniteScroller />
</Modal>
```

If so, then `InfiniteScroller` would just automatically work. If not… then you can maybe do something like this?

```
<Modal trigger={myTrigger}>
  <Conversation>
    <ThreadList ref={node => this.threadNode = node}>
      <InfiniteScroll scrollParent={this.threadNode} />
    </ThreadList>
  </Conversation>
</Modal>
```